tsdb/chunkenc/histogram_meta.go

```diff
@@ -140,6 +175,52 @@ func readZeroThreshold(br *bstreamReader) (float64, error) {
 	}
 }
 
+// isWholeWhenMultiplied checks to see if the number when multiplied by 1000 can
+// be converted into an integer without losing precision.
+func isWholeWhenMultiplied(in float64) bool {
+	i := uint(math.Round(in * 1000))
+	out := float64(i) / 1000
+	return in == out
+}
+
+// putCustomBound writes the custom bound to the bstream. It stores values from 0 to
+// 16.382 (inclusive) that are multiples of 0.001 in an unsigned var int of up to 2 bytes,
+// but needs 1 bit + 8 bytes for other values like negative numbers, numbers greater than
+// 16.382, or numbers that are not a multiple of 0.001, on the assumption that they are
+// less common. In detail:
+//   - Multiply the bound by 1000, without rounding.
+//   - If the multiplied bound is >= 0, <= 16382 and a whole number, store it as an
+//     unsigned var int.
+//   - Otherwise, store 0 as an unsigned var int, followed by the 8 bytes of the original
+//     bound as a float64.
+func putCustomBound(b *bstream, f float64) {
+	tf := f * 1000
+	if tf < 0 || tf > 16382 || !isWholeWhenMultiplied(f) {
+		b.putUvarint(0)
```

Comment:
nit: optimization: technically we could get away with writing a single 0 bit. But it gets a little complicated, so I'd do that in a subsequent PR.

Reply:
noted, let's leave the optimisations for a later PR as suggested

Reply:
Actually my idea doesn't work, because if we store the bucket bound 0.01, that becomes int 10, which is stored on 1 byte where the msb is 0, so it would not be possible to distinguish between floats and single byte integers , unless we waste a bit for every integer. So the modified algorithm would be: if number*1000 is whole then write bit 1 and then the uvarint, if not then write bit 0 and the float.